File: ch7/7-8.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const pageRows = 25

func main() {
	var pages, rows int
	if len(os.Args) == 1 {
		fmt.Printf("Not enough arguments!\n")
		os.Exit(1)
	}
	for i, n := range os.Args {
		if i == 0 {
			continue
		}
		f, err := os.Open(n)
		if err != nil {
			fmt.Printf("%s\n", err)
			os.Exit(1)
		}
		pages = countPages(f)
		fmt.Printf("File: %s, %d", n, pages)
		if pages > 1 {
			fmt.Printf(" pages")
		} else {
			fmt.Printf(" page")
		}
		fmt.Println()
		rows = 0
		_, err = f.Seek(0, 0)
		if err != nil {
			fmt.Printf("%s\n", err)
			os.Exit(1)
		}
		s := bufio.NewScanner(f)
		for s.Scan() {
			line := s.Text()
			rows++
			fmt.Printf("%s\n", line)
		}
		if err := s.Err(); err != nil {
			fmt.Printf("%s\n", err)
			os.Exit(1)
		}
		for i := rows % pageRows; i <= pageRows; i++ {
			fmt.Println()
		}
		f.Close()
	}
}

func countPages(f *os.File) int {
	var rows int
	s := bufio.NewScanner(f)
	for s.Scan() {
		rows++
	}
	if err := s.Err(); err != nil {
		fmt.Printf("%s\n", err)
		os.Exit(1)
	}
	return int(rows/pageRows + 1)
}

File: kr/ch7/7-8.c
#include <stdio.h>
#include <stdlib.h>

#define MAXLINE  1024
#define PAGEROWS 50

int count_pages(FILE *f);

int main(int argc, char *argv[])
{
    FILE *f;
    int pages = 0, rows, i;
    char line[MAXLINE];

    if (argc == 1) {
        printf("Not enough arguments!\n");
        return 0;
    }

    while (--argc > 0) {
        if ((f = fopen(*++argv, "r")) != NULL) {
            pages = count_pages(f);

            printf("File: %s, %d", *argv, pages);

            if (pages > 1) {
                printf(" pages");
            } else {
                printf(" page");
            }

            printf("\n");
            rows = 0;

            if ((fgets(line, MAXLINE, f)) == NULL) {
                printf("NULL");
            }

            while ((fgets(line, MAXLINE, f)) != NULL) {
                rows++;
                printf("%s", line);
            }

            for (i = rows % PAGEROWS; i <= PAGEROWS; i++) {
                printf("\n");
            }

            fclose(f);
        } else {
            fprintf(stderr, "can't open %s", *argv);
            exit(1);
        }
    }

    return 0;
}

int count_pages(FILE *f)
{
    int rows = 0;
    char line[MAXLINE];

    while ((fgets(line, MAXLINE, f)) != NULL) {
        rows++;
    }

    rewind(f);

    return rows / PAGEROWS + 1;
}
